src/nbl/asset/utils/IShaderCompiler.cpp

```diff
@@ -22,6 +25,73 @@ IShaderCompiler::IShaderCompiler(core::smart_refctd_ptr<system::ISystem>&& syste
     m_defaultIncludeFinder = core::make_smart_refctd_ptr<CIncludeFinder>(core::smart_refctd_ptr(m_system));
 }
 
+static void* SzAlloc(ISzAllocPtr p, size_t size) { p = p; return _NBL_ALIGNED_MALLOC(size, _NBL_SIMD_ALIGNMENT); }
+static void SzFree(ISzAllocPtr p, void* address) { p = p; _NBL_ALIGNED_FREE(address); }
+
+inline core::smart_refctd_ptr<ICPUShader> nbl::asset::IShaderCompiler::compileToSPIRV(const std::string_view code, const SCompilerOptions& options) const
+{
+    CCache::SEntry entry;
+    std::vector<CCache::SEntry::SPreprocessingDependency> dependencies;
+    if (options.readCache || options.writeCache)
+        entry = std::move(CCache::SEntry(code, options));
+
+    if (options.readCache)
+    {
+        auto found = options.readCache->find_impl(entry, options.preprocessorOptions.includeFinder);
+        if (found != options.readCache->m_container.end())
+        {
+            if (options.writeCache)
+            {
+                CCache::SEntry writeEntry = *found;
+                options.writeCache->insert(std::move(writeEntry));
+            }
+            return found->decodeShader();
+        }
+    }
+
+    auto retVal = compileToSPIRV_impl(code, options, options.writeCache ? &dependencies : nullptr);
+    // compute the SPIR-V shader content hash
+    {
+        auto backingBuffer = retVal->getContent();
+        const_cast<ICPUBuffer*>(backingBuffer)->setContentHash(backingBuffer->computeContentHash());
+    }
+
+    std::cout << "writeCache: " << options.writeCache << "\n";
+    if (options.writeCache)
+    {
+        auto* spirvBuffer = retVal->getContent();
+        size_t propsSize = LZMA_PROPS_SIZE;
+        size_t destLen = spirvBuffer->getSize() + spirvBuffer->getSize() / 3 + 128;
+        std::vector<unsigned char> compressedSpirv = {};
+        compressedSpirv.resize(propsSize + destLen);
+
+        CLzmaEncProps props;
+        LzmaEncProps_Init(&props);
+        props.dictSize = 1 << 16; // 64KB
+        props.writeEndMark = 1;
+
+        ISzAlloc alloc = { SzAlloc, SzFree };
+        int res = LzmaEncode(
+            compressedSpirv.data() + LZMA_PROPS_SIZE, &destLen,
+            reinterpret_cast<const unsigned char*>(spirvBuffer->getPointer()), spirvBuffer->getSize(),
+            &props, compressedSpirv.data(), &propsSize, props.writeEndMark,
+            nullptr, &alloc, &alloc);
+
+        assert(propsSize == LZMA_PROPS_SIZE);
+        assert(res == SZ_OK);
+
+        auto compressedSpirvBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(propsSize + destLen);
+        memcpy(compressedSpirvBuffer->getPointer(), compressedSpirv.data(), compressedSpirvBuffer->getSize());
+
+        entry.dependencies = std::move(dependencies);
+        entry.spirv = std::move(compressedSpirvBuffer);
+        entry.uncompressedSize = spirvBuffer->getSize();
```

Comment:
maybe `SEntry` needs different constructors:
- for a half-empty dummy lookup element
- for creating a complete entry from deserialization
- for making a complete entry to insert into write cache
and this needs to move into the latter

Reply:
> for creating a complete entry from deserialization
is it even doable with nlohmann_json?

Reply:
my comment is mostly about not having the LZMA code sit in `IShaderCompiler` methods but rather in `SEntry`